uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php

```diff
@@ -847,6 +847,76 @@ public function testFindOverlappingLeaveRequestsForOneOverlappingLeaveRequest()
     $this->assertEquals($leaveRequest1->id, $overlappingRequests[0]->id);
   }
 
+  public function testManagerCanCancelOrRejectUserLeaveRequestEvenIfBalanceChangeIsGreaterThanPeriodEntitlementBalanceChangeWhenAllowOveruseFalse() {
+    $manager = ContactFabricator::fabricate();
+    $staff = ContactFabricator::fabricate();
+    $entitlementFromDate = CRM_Utils_Date::processDate('2016-01-01');
+    $entitlementToDate = CRM_Utils_Date::processDate('2016-12-31');
+    $requestFromDate = new DateTime('2016-06-14');
+    $requestToDate = new DateTime('2016-06-14');
+    $requestFromDateType = $this->leaveRequestDayTypes['all_day']['value'];
+    $requestToDateType = $this->leaveRequestDayTypes['all_day']['value'];
+
+    $this->registerCurrentLoggedInContactInSession($manager['id']);
+    $this->setContactAsLeaveApproverOf($manager, $staff, null, null, true, 'has Leaves Approved By');
+
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => $entitlementFromDate,
+      'end_date'   => $entitlementToDate,
+    ]);
+
+    $absenceType = AbsenceTypeFabricator::fabricate([
+      'allow_overuse' => 0
+    ]);
+
+    $periodEntitlement = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $absenceType->id,
+      'contact_id' => $staff['id'],
+      'period_id' => $absencePeriod->id
+    ]);
+
+    $entitlementBalanceChange = 0;
+    $this->createLeaveBalanceChange($periodEntitlement->id, $entitlementBalanceChange);
+    $periodStartDate = $absencePeriod->start_date;
+
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $periodEntitlement->contact_id],
+      ['period_start_date' => $periodStartDate]
+    );
+
+    $workPattern = WorkPatternFabricator::fabricateWithA40HourWorkWeek();
+    ContactWorkPatternFabricator::fabricate([
+      'contact_id' => $periodEntitlement->contact_id,
+      'pattern_id' => $workPattern->id
+    ]);
+
+    $leaveRequestStatuses = array_flip(LeaveRequest::buildOptions('status_id', 'validate'));
+
+    $leaveRequestData = [
+      'type_id' => $absenceType->id,
+      'contact_id' => $periodEntitlement->contact_id,
+      'status_id' => $leaveRequestStatuses['awaiting_approval'],
+      'from_date' => $requestFromDate->format('YmdHis'),
+      'from_date_type' => $requestFromDateType,
+      'to_date' => $requestToDate->format('YmdHis'),
+      'to_date_type' => $requestToDateType,
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE
+    ];
+
+    $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation($leaveRequestData);
+
+    $updatedLeaveRequestData = $leaveRequestData;
+    $updatedLeaveRequestData['id'] = $leaveRequest->id;
+
+    foreach (array("cancelled", "rejected") as $statusName) {
+      $requestNewStatus = $leaveRequestStatuses[$statusName];
+      $updatedLeaveRequestData['status_id'] = $requestNewStatus;
+      $cancelledLeaveRequest = LeaveRequest::create($updatedLeaveRequestData);
```

Comment:
I'm just having some afterthoughts here. The test demonstrates canceling a previously created leave request. What is the leave request is already approved and something changes e.g the status, will an error be thrown because I think there's no where in the code where we check if a leave request is already approved before comparing the leave balance to the entitlement balance.
Can you add a simple test for this scenario?

Reply:
Do you mean the following use case:
1) Leave request approved when the balance was enough
2) Some time has gone and the balance is now 0
3) Manager can now Reject/Cancel it but cannot put back to Approved?

Reply:
(1) Leave request approved when balance is enough.
(2) Leave request balance is Zero
(3) Manager changes date or status
(4)On trying to do that, will the code here fail? https://github.com/civicrm/civihr/blob/d85d757f9fa6b34eb6f915a47cf5052f3bad2861/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php#L462-L462
(5) Reason is the already approved leave request is calculated as part of the entitlement balance.
E.g Entitlement is 1, Leave Request initially  = -1
Entitlement is now 0
Manager updates status/dates for Approved Leave.
We now compare if `$leaveRequestBalance > $currentBalance `
i.e 1 > 0
Meanwhile it shouldn't be so because the approved leave has already been deducted.
Similar to what we did for TOIL here https://github.com/civicrm/civihr/blob/d85d757f9fa6b34eb6f915a47cf5052f3bad2861/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php#L293-L293

Reply:
Such test passes.